hardwareModules/RoFICoM/software/i2c_test/vl53l1_lidar/lidar.hpp

```diff
@@ -0,0 +1,130 @@
+        , _deviceAddress( deviceAddress )
+        , _communicationSpeed( commSpeed )
+    {
+
+    }
+
+    result_type initialize()
+    {
+        VL53L1_Error status;
+
+        status = VL53L1_platform_init( &_device, _deviceAddress, VL53L1_I2C, _communicationSpeed );
+        if (status != VL53L1_ERROR_NONE) {
+            return result_type::error( errorToString( status ) );
+        }
+
+        status = VL53L1_WaitDeviceBooted( &_device );
+        if ( status != VL53L1_ERROR_NONE ) {
+            return result_type::error( errorToString( status ) );
+        }
+
+        status = VL53L1_DataInit( &_device );
+        if (status != VL53L1_ERROR_NONE) {
+            return result_type::error( errorToString( status ) );
+        }
+
+        status = VL53L1_StaticInit( &_device );
+        if (status != VL53L1_ERROR_NONE) {
+            return result_type::error( errorToString( status ) );
+        }
+
+        return atoms::make_result_value< Void >();
+    }
+
+    result_type startMeasurement()
+    {
+        VL53L1_Error status = VL53L1_StartMeasurement( &_device );
+        if (status != VL53L1_ERROR_NONE) {
+            return result_type::error( errorToString( status ) );
+        }
+
+        return atoms::make_result_value< Void >();
+    }
+
+    result_type stopMeasurement()
+    {
+        VL53L1_Error status = VL53L1_ClearInterruptAndStartMeasurement( &_device );
+        if ( status != VL53L1_ERROR_NONE ) {
+            return result_type::error( errorToString( status ) );
+        }
+
+        status = VL53L1_StopMeasurement( &_device );
+        if ( status != VL53L1_ERROR_NONE ) {
+            return result_type::error( errorToString( status ) );
+        }
+
+        return atoms::make_result_value< Void >();
+    }
+
+    result_type waitMeasurementDataReady()
+    {
+        VL53L1_Error status = VL53L1_WaitMeasurementDataReady( &_device );
+        if ( status != VL53L1_ERROR_NONE ) {
+            return result_type::error( errorToString( status ) );
+        }
+
+        return atoms::make_result_value< Void >();
+    }
+
+    result_type clearInterruptAndStartMeasurement()
+    {
+        VL53L1_Error status = VL53L1_ClearInterruptAndStartMeasurement( &_device );
+        if ( status != VL53L1_ERROR_NONE ) {
+            return result_type::error( errorToString( status ) );
+        }
+
+        return atoms::make_result_value< Void >();
+    }
+
+    Result< VL53L1_RangingMeasurementData_t, std::string> getRangingMeasurementData()
+    {
+        VL53L1_RangingMeasurementData_t rangingMeasurementData;
+
+        VL53L1_Error status = VL53L1_GetRangingMeasurementData( &_device, &rangingMeasurementData );
+        if ( status != VL53L1_ERROR_NONE ) {
+            return atoms::result_error( errorToString( status ) );
+        }
+
+        return atoms::result_value< VL53L1_RangingMeasurementData_t >( rangingMeasurementData );
+    }
+
+private:
+    std::string errorToString( VL53L1_Error err )
```

Comment:
Celkově by bylo rozumné se `std::string` jako takovému vyhnout. Má poměrně velký overhead (zejména co se velikosti kódu týče) a zrovna na chyby by mohl stačit `std::string_view`

Reply:
Problém je, že ta knihovna kopíruje chybovou hlášku da daného bufferu, takže by to `std::string_view` nemohl být, protože úložiště není konstantní. Pokud by buffer, který tomu dávám by byl `static`, ale to by znamenalo, že chybová zpráva, by se mohla změnit, než by byla přečtena (kdyby proběhlo zavolání jiné funkce volající `errorToString`, před přečtení chyby). Jestli teda chápu `std::string_view` dobře.

Reply:
Všechen převod knihovní chyby používá následující buffer, který je funkci předán od uživatele knihovny: https://github.com/paradise-fi/RoFI/blob/093636ca9ef797d6290f64299a913262c063e0e0/hardwareModules/RoFICoM/software/i2c_test/vl53l1_lidar/core/inc/vl53l1_ll_def.h#L1089-L1090